_projects/coala_language_server.md

```diff
@@ -42,17 +42,25 @@ a rapidly growing list of editor and language communities:
 We have a state-of-art implementation in
 [coala-vs-code](https://github.com/coala/coala-vs-code) but it is designed
 for VS Code only. We need to design and implement a general mechanism for
-multiple editors/IDEs support.
+multiple editors/IDEs support. We also need clients written to support coala
+langserver for most if not all the supported editors. This is not because of
```

Comment:
Some editors like VSCode with programmable server interface declarations can support TCP interfaces, but using an stdio stream is preferred by most of implementations to improve efficiency. Fortunately, most plugins support configuring the client via a simple configuration file.

Based on https://insights.stackoverflow.com/survey/2018#technology-most-popular-development-environments and the existing implementation, testing coala language server with VSCode would be the best. The end product sould also be tested with other configurable clients for example https://github.com/prabirshrestha/vim-lsp.

The clients developed should be shareable with other language servers, pyls supports both a TCP and stdio mode and should be completely compatible.

As the server is to be published as a python package, I think it would be wise to have the server independent of the client and thus have separate repositories for both. I also think that it would be safe to initially have a unified repository for clients and branch out to multiple repositories when required.

Reply:
Generally, agree with @ksdme.

Reply:
We're working on splitting the JSON RPC server out of PYLS. 

On the client side, some editors can only support language servers through the means of extensions. So it's common in VSCode for example to publish a new client for each server. One that exposes the various configuration options and supported command IDs. The PYLS VSCode client (included in the repo but not published) is simply the example language client project plus a package.json that contains the configuration options. For other editors, such as Vim, it's sufficient just to declare a command line that supports the LSP over stdio.